src/persist-client/src/usage.rs

```diff
@@ -11,109 +11,373 @@
+            client.cfg,
+            Arc::clone(&client.consensus),
+            Arc::clone(&client.blob),
+            Arc::clone(&client.metrics),
+        ));
+        StorageUsageClient {
+            blob: client.blob,
+            metrics: client.metrics,
+            state_versions,
+        }
     }
 
-    /// Returns the size (in bytes) of all blobs owned by a given [ShardId]
-    pub async fn shard_size(&self, shard_id: &ShardId) -> u64 {
-        retry_external(
-            &self.metrics.retries.external.storage_usage_shard_size,
-            || self.size(BlobKeyPrefix::Shard(shard_id)),
-        )
-        .await
+    /// Computes [ShardUsage] for a single shard.
+    pub async fn shard_usage(&self, shard_id: ShardId) -> ShardUsage {
+        let mut blob_usage = self.blob_raw_usage(BlobKeyPrefix::Shard(&shard_id)).await;
+        let blob_usage = blob_usage.by_shard.remove(&shard_id).unwrap_or_default();
+        self.shard_usage_given_blob_usage(shard_id, &blob_usage)
+            .await
     }
 
     /// Returns a map of [ShardId] to its size (in bytes) stored in persist, as
     /// well as the total bytes stored for unattributable data (the map key is
     /// None) where the shard id is unknown. This latter count _should_ always
     /// be 0, so it being nonzero indicates either persist wrote an invalid blob
     /// key, or another process is storing data under the same path (!)
+    ///
+    /// TODO: Remove this in favor of using [Self::shards_usage] directly.
     pub async fn shard_sizes(&self) -> BTreeMap<Option<ShardId>, u64> {
+        let usage = self.shards_usage().await;
+        let mut ret = usage
+            .by_shard
+            .iter()
+            .map(|(id, usage)| (Some(*id), usage.total_bytes()))
+            .collect::<BTreeMap<_, _>>();
+        if usage.unattributable_bytes > 0 {
+            ret.insert(None, usage.unattributable_bytes);
+        }
+        ret
+    }
+
+    /// Computes [ShardUsage] for every shard in an env.
+    pub async fn shards_usage(&self) -> ShardsUsage {
+        let blob_usage = self.blob_raw_usage(BlobKeyPrefix::All).await;
+        self.metrics
+            .audit
+            .blob_batch_part_bytes
+            .set(blob_usage.batch_part_bytes);
+        self.metrics
+            .audit
+            .blob_batch_part_count
+            .set(blob_usage.batch_part_count);
+        self.metrics
+            .audit
+            .blob_rollup_bytes
+            .set(blob_usage.rollup_size);
+        self.metrics
+            .audit
+            .blob_rollup_count
+            .set(blob_usage.rollup_count);
+        self.metrics.audit.blob_bytes.set(blob_usage.total_size);
+        self.metrics.audit.blob_count.set(blob_usage.total_count);
+
+        let mut shards_usage = ShardsUsage {
+            by_shard: BTreeMap::new(),
+            unattributable_bytes: blob_usage.unattributable_bytes,
+        };
+        // TODO: FuturesUnordered with some concurrency limit?
```

Comment:
The `test_storage_usage_collection_interval_timestamps` flake spooked me, so went ahead and did this TODO now